File: scala/P11.scala
// P11 Modified run-length encoding.

// List of objects
def f1[T](list: List[(Int, T)]): List[Any] = list.map {
  case (1, x) => x
  case it => it
}

// With custom data
abstract sealed class Entry[T]
case class Value[T](v: T) extends Entry[T]
case class Sequence[T](n: Int, v: T) extends Entry[T]

def f2[T](list: List[(Int, T)]): List[Entry[T]] = list.map {
  case (1, x) => Value(x)
  case (n, x) => Sequence(n, x)
}

File: haskell/p11.hs
-- P11 Modified run-length encoding.

module P11 where

data Entry a = Value a | Sequence Int a deriving Show

f1 :: [(Int, a)] -> [Entry a]
f1 = map toEntry
    where toEntry (1, x) = Value x
          toEntry (n, x) = Sequence n x